File: packages/frontend/src/store/action/message.ts
import { createAsyncThunk } from "@reduxjs/toolkit";
import { generateTemplateId, WS_EVENT } from "core";
import { formatMessage } from "@/utils/formatMessage";
import { normalizeMessage } from "@/Api";
import {
  addMessage,
  replaceMessage,
  scrollToEnd,
  updateMessageStatus,
  updateReplyMessage,
} from "../reducer/room";
import { IMessage } from "@/interfaces";
import { topUserRoom, updateUserRoomReadSeq } from "../reducer/user";

const { toast } = createStandaloneToast();

const sendOptimisticMessage = async (
  data: Partial<IMessage>,
  dispatch: any,
  getState: any,
  options?: { localMessageId?: string; skipAdd?: boolean }
) => {
  const state: any = getState();
  const room = state?.room?.data;
  const user = state?.user?.data;
  const replyMessage = data.reply ?? state?.room?.replyMessage;

  if (!options?.localMessageId && state?.room?.replyMessage?.id) {
    Object.assign(data, {
      replyId: state?.room?.replyMessage?.id,
    });
  }

  const localMessageId =
    options?.localMessageId ?? `local-${generateTemplateId()}`;
  const lastSeq = room?.message?.[room.message.length - 1]?.seq ?? 0;
  const optimisticMessage: IMessage = normalizeMessage({
    id: localMessageId,
    seq: data.seq ?? lastSeq + 1,
    contentType: data.contentType,
    channelId: data.channelId,
    roomId: data.channelId,
    textMessage: data.textMessage,
    mediaMessage: data.mediaMessage,
    userId: data.userId,
    replyId: data.replyId,
    reply: replyMessage,
    user,
    createdAt: data.createdAt ?? new Date(),
    updatedAt: new Date(),
    isRecalled: false,
    localStatus: "sending",
  });

  if (options?.skipAdd) {
    dispatch(
      replaceMessage({
        id: localMessageId,
        message: optimisticMessage,
      })
    );
  } else {
    dispatch(addMessage(optimisticMessage));
    dispatch(topUserRoom(optimisticMessage));
    dispatch(scrollToEnd(false));
    dispatch(updateReplyMessage());
  }

  try {
    const formatMsg = await formatMessage(data);
    const wsRes = await ws.sendMsgPromise<IMessage>(formatMsg);
    if (wsRes.code === 1) {
      throw new Error(wsRes.message);
    }
    const msg = normalizeMessage(wsRes.data);
    dispatch(replaceMessage({ id: localMessageId, message: msg }));
    dispatch(topUserRoom(msg));
    dispatch(
      updateUserRoomReadSeq({
        channelId: msg.channelId,
        readMessage: {
          lastReadSeq: NaN,
          operator: msg.userId,
          readSeq: {
            [msg.userId]: msg.seq,
          },
        },
      })
    );
  } catch (error: any) {
    dispatch(
      updateMessageStatus({
        id: localMessageId,
        localStatus: "failed",
      })
    );
    toast({
      description: error?.message ?? "Failed to send message",
      status: "error",
      position: "top",
      duration: 1000,
    });
    return Promise.reject(error?.message ?? "Failed to send message");
  }
};

// 发送一条新消息
export const sendMessageAction = createAsyncThunk<void, Partial<IMessage>>(
  `sendMessage`,
  async (data, { dispatch, getState }) => {
    return sendOptimisticMessage(data, dispatch, getState);
  }
);

export const retryMessageAction = createAsyncThunk<void, IMessage>(
  `retryMessage`,
  async (message, { dispatch, getState, rejectWithValue }) => {
    if (!message.id || !message.channelId || !message.userId) {
      return rejectWithValue("Missing message retry payload");
    }

    try {
      await sendOptimisticMessage(message, dispatch, getState, {
        localMessageId: String(message.id),
        skipAdd: true,
      });
    } catch (error: any) {
      return rejectWithValue(
        error?.message ?? error ?? "Failed to retry message"
      );
    }
  }
);

export const markReadMessageThunk = createAsyncThunk<void, Partial<IMessage>>(
  `markReadMessage`,
  async (message) => {
    if (!message.channelId) return;
    const readMessage: Partial<IMessage> = {
      ...message,
      contentType: "READ_MESSAGE",
    };
    ws.sendMsg<IMessage>(readMessage, WS_EVENT.SEND_MSG);
  }
);

export const recallMessageThunk = createAsyncThunk<void, Partial<IMessage>>(
  `recallMessage`,
  async (message) => {
    if (!message.channelId) return;
    const recallMessage: Partial<IMessage> = {
      ...message,
      contentType: "RECALL_MESSAGE",
    };
    ws.sendMsg<IMessage>(recallMessage, WS_EVENT.SEND_MSG);
  }
);

File: packages/backend-go/internal/model/message.go
package model

import (
	"gorm.io/gorm"
)

type Message struct {
	gorm.Model
	ID            uint   `gorm:"primarykey" json:"id"`
	Seq           int    `gorm:"column:seq;index" json:"seq"`
	ContentType   string `gorm:"column:content_type" json:"content_type"`
	ChannelId     string `gorm:"column:channel_id;index" json:"channel_id"`
	TextMessage   string `gorm:"column:text_message;type:json" json:"text_message"`
	MediaMessage  string `gorm:"column:media_message;type:json" json:"media_message"`
	ReadMessage   string `gorm:"column:read_message;type:json" json:"read_message"`
	RecallMessage string `gorm:"column:recall_message;type:json" json:"recall_message"`
	SystemMessage string `gorm:"column:system_message;type:json" json:"system_message"`
	UserId        string `gorm:"column:user_id;index" json:"user_id"`
	RoomId        string `gorm:"column:room_id;index" json:"room_id"`
	ReplyId       string `gorm:"column:reply_id" json:"reply_id"`
	IsRecalled    bool   `gorm:"column:is_recalled;default:false" json:"is_recalled"`
}

func (m *Message) normalizeJSONFields() {
	if m.TextMessage == "" {
		m.TextMessage = "null"
	}
	if m.MediaMessage == "" {
		m.MediaMessage = "null"
	}
	if m.ReadMessage == "" {
		m.ReadMessage = "null"
	}
	if m.RecallMessage == "" {
		m.RecallMessage = "null"
	}
	if m.SystemMessage == "" {
		m.SystemMessage = "null"
	}
}

func (m *Message) BeforeCreate(tx *gorm.DB) error {
	m.normalizeJSONFields()
	return nil
}

func (m *Message) BeforeSave(tx *gorm.DB) error {
	m.normalizeJSONFields()
	return nil
}

func (Message) TableName() string {
	return "messages"
}
